keanu-project/src/main/java/io/improbable/keanu/vertices/generic/probabilistic/discrete/CategoricalVertex.java

```diff
@@ -21,50 +26,55 @@
 import static java.util.stream.Collectors.toMap;
 
 
-public class CategoricalVertex<T> extends Vertex<Tensor<T>> implements Probabilistic<Tensor<T>> {
+public class CategoricalVertex<T, TENSOR extends Tensor<T>> extends Vertex<TENSOR> implements Probabilistic<TENSOR> {
```

Comment:
I wanted to avoid doing this, because I'm not a big fan of forcing people to specify both the element type and the tensor type when instantiating a categorical vertex. But I guess there is no other way...

Reply:
I think this is necessary but I agree it's ugly and not something the user should care about. What we can do is define a Typed class that has T and TENSOR filled in. `e.g. CategoricalIntegerVertex extends CategoricalVertex<Integer, IntegerTensor>`.